src/librbd/ManagedLock.cc

```diff
@@ -0,0 +1,501 @@
+  m_state = STATE_ACQUIRING;
+
+  assert(m_watcher->is_registered());
+  uint64_t watch_handle = m_watcher->get_watch_handle();
+
+  m_cookie = ManagedLock<W>::encode_lock_cookie(watch_handle);
+  AcquireRequest<W>* req = AcquireRequest<W>::create(this,
+    util::create_context_callback<
+      ManagedLock<W>, &ManagedLock<W>::handle_acquire_lock>(this));
+  m_work_queue->queue(new C_SendRequest<AcquireRequest<W>>(req), 0);
+}
+
+template <typename W>
+void ManagedLock<W>::handle_acquire_lock(int r) {
+  ldout(m_cct, 10) << this << " " << __func__ << ": r=" << r << dendl;
+
+  if (r == -EBUSY || r == -EAGAIN) {
+    ldout(m_cct, 5) << "unable to acquire exclusive lock" << dendl;
+  } else if (r < 0) {
+    lderr(m_cct) << "failed to acquire exclusive lock:" << cpp_strerror(r)
+               << dendl;
+  } else {
+    ldout(m_cct, 5) << "successfully acquired exclusive lock" << dendl;
+  }
+
+  State next_state = (r < 0 ? STATE_UNLOCKED : STATE_LOCKED);
+  if (r == -EAGAIN) {
+    r = 0;
+  }
+
+  Mutex::Locker locker(m_lock);
+  complete_active_action(next_state, r);
+}
+
+template <typename W>
+void ManagedLock<W>::send_reacquire_lock() {
+  assert(m_lock.is_locked());
+
+  if (m_state != STATE_LOCKED) {
+    complete_active_action(m_state, 0);
+    return;
+  }
+
+  uint64_t watch_handle = m_watcher->get_watch_handle();
+  if (watch_handle == 0) {
+     // watch (re)failed while recovering
+     lderr(m_cct) << this << " " << __func__ << ": "
+                  << "aborting reacquire due to invalid watch handle" << dendl;
+     complete_active_action(STATE_LOCKED, 0);
+     return;
+  }
+
+  m_new_cookie = ManagedLock<W>::encode_lock_cookie(watch_handle);
+  if (m_cookie == m_new_cookie) {
+    ldout(m_cct, 10) << this << " " << __func__ << ": "
+                   << "skipping reacquire since cookie still valid" << dendl;
+    complete_active_action(STATE_LOCKED, 0);
+    return;
+  }
+
+  ldout(m_cct, 10) << this << " " << __func__ << dendl;
+  m_state = STATE_REACQUIRING;
+
+  ReacquireRequest<W>* req = ReacquireRequest<W>::create(this, m_new_cookie,
+    util::create_context_callback<
+      ManagedLock, &ManagedLock<W>::handle_reacquire_lock>(this));
+  req->send();
+}
+
+template <typename W>
+void ManagedLock<W>::handle_reacquire_lock(int r) {
+  Mutex::Locker locker(m_lock);
+
+  ldout(m_cct, 10) << this << " " << __func__ << ": r=" << r << dendl;
+
+  assert(m_state == STATE_REACQUIRING);
+  if (r < 0) {
+    if (r == -EOPNOTSUPP) {
+      ldout(m_cct, 10) << this << " " << __func__ << ": "
+                     << "updating lock is not supported" << dendl;
+    } else {
+      lderr(m_cct) << this << " " << __func__ << ": "
+                 << "failed to update lock cookie: " << cpp_strerror(r)
+                 << dendl;
+    }
+
+    if (!is_shutdown_locked()) {
+      // queue a release and re-acquire of the lock since cookie cannot
+      // be updated on older OSDs
+      execute_action(ACTION_RELEASE_LOCK, nullptr);
```

Comment:
This needs to happen at the upper layer (ExclusiveLock) since the release step involves shutting down the journaler / object map. If you used ManagedLock as a parent class of ExclusiveLock, you could have the necessary hooks via virtual methods to start the release/reacquire process